helix-core/src/numbers.rs

```diff
@@ -0,0 +1,502 @@
+use std::borrow::Cow;
+
+use ropey::RopeSlice;
+
+use crate::{
+    textobject::{textobject_word, TextObject},
+    Range, Tendril,
+};
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct NumberIncrementor<'a> {
+    pub range: Range,
+    pub value: i64,
+    pub radix: u32,
+
+    text: RopeSlice<'a>,
+}
+
+impl<'a> NumberIncrementor<'a> {
+    /// Return information about number under rang if there is one.
+    pub fn from_range(text: RopeSlice, range: Range) -> Option<NumberIncrementor> {
+        // If the cursor is on the minus sign of a number we want to get the word textobject to the
+        // right of it.
+        let range = if range.to() < text.len_chars()
+            && range.to() - range.from() <= 1
+            && text.char(range.from()) == '-'
+        {
+            Range::new(range.from() + 1, range.to() + 1)
+        } else {
+            range
+        };
+
+        let range = textobject_word(text, range, TextObject::Inside, 1, false);
+
+        // If there is a minus sign to the left of the word object, we want to include it in the range.
+        let range = if range.from() > 0 && text.char(range.from() - 1) == '-' {
+            range.extend(range.from() - 1, range.from())
+        } else {
+            range
+        };
+
+        let word: String = text
+            .slice(range.from()..range.to())
+            .chars()
+            .filter(|&c| c != '_')
+            .collect();
+        let (radix, prefixed) = if word.starts_with("0x") {
+            (16, true)
+        } else if word.starts_with("0o") {
+            (8, true)
+        } else if word.starts_with("0b") {
+            (2, true)
+        } else {
+            (10, false)
+        };
+
+        let number = if prefixed { &word[2..] } else { &word };
+
+        let value = i128::from_str_radix(number, radix).ok()?;
+        if (value.is_positive() && value.leading_zeros() < 64)
+            || (value.is_negative() && value.leading_ones() < 64)
+        {
+            return None;
+        }
```

Comment:
I don't understand what does this do. Can you please put a comment here? Is it because the integer may be too large?

Reply:
In Vim's implementation, decrementing 0x0 results in 0xffffffffffffffff, which is -1 for i64. However, if we try to parse that with i64::from_str_radix we get an error because it tries to parse it as a positive number and overflows. If I parse it as an i128, I can cast it to i64 to get the negative value.

This check just makes sure the number fits in an i64. You may be wondering why I didn't use `try_from` to convert to i64, and the reason is it wouldn't work for parsing negative numbers that aren't base 10.

```Rust
let value = i128::from_str_radix("ffffffffffffffff", 16).unwrap();
let value = value as i64;
// `value` now is -1
```
but
```Rust
let value = i128::from_str_radix("ffffffffffffffff", 16).unwrap();
let value = i64::try_from(value).unwrap();
```
This version will get a `TryFromIntError` error because the original i128 is actually a positive number that's too large to fit in an i64.

Can you think of a better way to handle this, or do you think adding a comment is the best option?

Reply:
Maybe the best thing to do here is change the behavior from what Vim does.

Vim treats `0xFFFFFFFFFFFFFFFF` as `-1`, because that's the bit pattern of -1 for a 64 bit two-complement integer. Vim only handles the minus sign for base 10 numbers.

So in Vim `-0x01` gets incremented to `-0x02` instead of `0x00` because it ignores the minus sign. We could make helix only interpret the number as negative if it has a minus sign.

Also in Vim `0x00` gets decremented to `0xffffffffffffffff`, but we could have helix decrement it to `-0x01`.

But this would also mean that `0xffffffffffffffff` would be too large and wouldn't be parsed as an i64.

What do you think?

Reply:
I think the easiest way here might be just store it as i128 but either way, this is already merged so it should be good.